modules/library/main/src/main/java/org/geotools/renderer/crs/ProjectionHandler.java

```diff
@@ -343,29 +306,132 @@ private void addTransformedEnvelope(ReferencedEnvelope re, List<ReferencedEnvelo
+        CoordinateReferenceSystem crs = re.getCoordinateReferenceSystem();
+        boolean isLatLon = CRS.getAxisOrder(crs) == CRS.AxisOrder.NORTH_EAST;
+        double minX = isLatLon ? re.getMinY() : re.getMinX();
+        double maxX = isLatLon ? re.getMaxY() : re.getMaxX();
+        double minY = isLatLon ? re.getMinX() : re.getMinY();
+        double maxY = isLatLon ? re.getMaxX() : re.getMaxY();
+        double extent = maxX - minX;
+        List<ReferencedEnvelope> envelopesToBeAdded = new ArrayList<>();
+        if (extent > 360) {
+            // at least one whole world use case -> requested data covers the full world:
+            // let's set a -180,180 bbox.
+            // the wrapping projectionHandler and the gridCoverageReaders
+            // will do proper clones / intersections afterwards
+            minX = -180;
+            maxX = 180;
+            // Create a whole world envelope taking into account latLon/lonLat
+            ReferencedEnvelope envelope =
+                    new ReferencedEnvelope(
+                            isLatLon ? minY : minX,
+                            isLatLon ? maxY : maxX,
+                            isLatLon ? minX : minY,
+                            isLatLon ? maxX : maxY,
+                            crs);
+            envelopesToBeAdded.add(envelope);
+        } else {
+            // let's do some adjustments to "shift" the request around -180, 180 interval:
+            // we basically add or subtract 360° N times
+            // 1) let's count how many halfCircles (a 180° span) we are away from the zero
+
+            // 2) add/subtract 360° N times to move forward/backward the request, also
+            // keeping into account whether we are crossing the dateline or not
+            // (halfCircles % 2)
+
+            // 3) add/subtract the original extent to get the other value of the interval
+            // in order to move the whole window (Note that the extent won't be > 360°
+            // since we are inside the "else")
+
+            int halfCircles = 0;
+            if (minX < -180) {
+                halfCircles = (int) ((Math.abs(minX) / 180));
+                minX += (360 * ((halfCircles / 2) + (halfCircles % 2)));
+                maxX = minX + extent;
+            } else if (minX > 180) {
+                halfCircles = (int) (minX / 180);
+                minX -= (360 * ((halfCircles / 2) + (halfCircles % 2)));
+                maxX = minX + extent;
+            } else if (maxX < -180) {
+                halfCircles = (int) (Math.abs(maxX) / 180);
+                maxX += (360 * ((halfCircles / 2) + (halfCircles % 2)));
+                minX = maxX - extent;
+            } else if (maxX > 180) {
+                halfCircles = (int) (Math.abs(maxX) / 180);
+                maxX -= (360 * ((halfCircles / 2) + (halfCircles % 2)));
+                minX = maxX - extent;
+            }
+
+            if ((int) (minX / 180) < (int) (maxX / 180)) {
```

Comment:
Isn't this triggering every time (maxX - minX) > 180? Doesn't mean the dateline has been crossed though.

Reply:
Same as above, add comments.